packages/blockchain-link-utils/src/solana.ts

```diff
@@ -0,0 +1,398 @@
+        .map(ak => {
+            const targetAddress = ak.pubkey.toString();
+            const balanceDiff = extractAccountBalanceDiff(transaction, targetAddress);
+
+            // WSOL Transfers are counted as SOL transfers in the transaction effects, leading to duplicate
+            // entries in the tx history. This serves to filter out the WSOL transfers from the native effects.
+            if (wSolTransferInstruction && 'parsed' in wSolTransferInstruction) {
+                if (
+                    wSolTransferInstruction.parsed.info.destination === targetAddress ||
+                    wSolTransferInstruction.parsed.info.source === targetAddress
+                ) {
+                    return null;
+                }
+            }
+
+            if (!balanceDiff) {
+                return null;
+            }
+
+            return {
+                address: targetAddress,
+                amount: balanceDiff.postBalance.minus(balanceDiff.preBalance),
+            };
+        })
+        .filter((effect): effect is TransactionEffect => !!effect)
+        .filter(({ amount }) => !amount.isZero()); // filter out zero effects
+}
+
+export const getTargets = (
+    effects: TransactionEffect[],
+    txType: Transaction['type'],
+    accountAddress: string,
+): Transaction['targets'] =>
+    effects
+        .filter(effect => {
+            // for 'self` transaction there is only one effect
+            if (txType === 'self') {
+                return true;
+            }
+            // ignore all targets for unknown transactions
+            if (txType === 'unknown') {
+                return false;
+            }
+            // count in only positive effects, for `sent` tx they gonna be represented as negative, for `recv` as positive
+            return effect.amount.isGreaterThan(0);
+        })
+        .map((effect, i) => {
+            const target: Target = {
+                n: i,
+                addresses: [effect.address],
+                isAddress: true,
+                amount: effect.amount.abs().toString(),
+                isAccountTarget: effect.address === accountAddress && txType !== 'sent',
+            };
+            return target;
+        });
+
+const getTokenTransferTxType = (transfers: TokenTransfer[]) => {
+    if (transfers.every(({ type }) => type === 'recv')) {
+        return 'recv';
+    }
+
+    if (transfers.every(({ type }) => type === 'sent')) {
+        return 'sent';
+    }
+
+    return 'unknown';
+};
+
+const getNativeTransferTxType = (
+    effects: TransactionEffect[],
+    accountAddress: string,
+    transaction: ParsedTransactionWithMeta,
+) => {
+    if (
+        effects.length === 1 &&
+        effects[0]?.address === accountAddress &&
+        effects[0]?.amount.abs().isEqualTo(new BigNumber(transaction.meta?.fee || 0))
+    ) {
+        return 'self';
+    }
+
+    const senders = effects.filter(({ amount }) => amount.isNegative());
```

Comment:
nit pic: senders and receivers look like candidates to use `arrayPartition` utility